### PR TITLE
feat: employ sparse key encoding for shard lookup

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR employs spase key encoding to encode primary keys to bytes for primary key to pk id look up.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable/merge_tree/partition.rs [hunk 1]
```diff
@@ -578,12 +593,16 @@ impl Inner {
         }
     }
 
-    fn find_key_in_shards(&self, primary_key: &[u8]) -> Option<PkId> {
+    fn find_key_in_shards(&self, short_key: &[u8]) -> Option<PkId> {
```
Comment (evenyag) [inline src/mito2/src/memtable/merge_tree/partition.rs:596]:
```suggestion
    fn find_key_in_shards(&self, sparse_key: &[u8]) -> Option<PkId> {
```

File: src/mito2/src/memtable/merge_tree/tree.rs [hunk 2]
```diff
@@ -324,6 +351,34 @@ impl MergeTree {
     }
 }
 
+struct FieldWithId {
+    field: SortField,
+    column_id: ColumnId,
+}
+
+struct ShortEncoder {
```
Comment (evenyag) [inline src/mito2/src/memtable/merge_tree/tree.rs:359]:
```suggestion
struct SparseEncoder {
```

File: src/mito2/src/memtable/merge_tree/partition.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/partition.rs:596]: > fn find_key_in_shards(&self, short_key: &[u8]

I think better to rename it back to `primary_key`

File: src/mito2/src/memtable/merge_tree/tree.rs [hunk 3]
```diff
@@ -69,6 +73,15 @@ impl MergeTree {
                 .map(|c| SortField::new(c.column_schema.data_type.clone()))
                 .collect(),
         );
+        let short_encoder = SparseEncoder {
```
Comment (evenyag) [inline src/mito2/src/memtable/merge_tree/tree.rs:76]:
```suggestion
        let sparse_encoder = SparseEncoder {
```

**Review by evenyag — APPROVED** [review]
Mostly LGTM

**Pull request merged**